pkg/reconciler/customrun/customrun_test.go

```diff
@@ -0,0 +1,307 @@
+	"github.com/tektoncd/pipeline/test/names"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"knative.dev/pkg/apis"
+	cminformer "knative.dev/pkg/configmap/informer"
+	pkgreconciler "knative.dev/pkg/reconciler"
+
+	"knative.dev/pkg/system"
+
+	_ "knative.dev/pkg/system/testing" // Setup system.Namespace()
+)
+
+func initializeRunControllerAssets(t *testing.T, d test.Data) (test.Assets, func()) {
+	ctx, _ := ttesting.SetupFakeContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	test.EnsureConfigurationConfigMapsExist(&d)
+	c, informers := test.SeedTestData(t, ctx, d)
+	configMapWatcher := cminformer.NewInformedWatcher(c.Kube, system.Namespace())
+	ctl := NewController()(ctx, configMapWatcher)
+	if err := configMapWatcher.Start(ctx.Done()); err != nil {
+		t.Fatalf("error starting configmap watcher: %v", err)
+	}
+
+	if la, ok := ctl.Reconciler.(pkgreconciler.LeaderAware); ok {
+		la.Promote(pkgreconciler.UniversalBucket(), func(pkgreconciler.Bucket, types.NamespacedName) {})
+	}
+
+	return test.Assets{
+		Logger:     logging.FromContext(ctx),
+		Controller: ctl,
+		Clients:    c,
+		Informers:  informers,
+		Recorder:   controller.GetEventRecorder(ctx).(*record.FakeRecorder),
+		Ctx:        ctx,
+	}, cancel
+}
+
+func getRunName(run v1beta1.CustomRun) string {
+	return strings.Join([]string{run.Namespace, run.Name}, "/")
+}
+
+// getRunController returns an instance of the TaskRun controller/reconciler that has been seeded with
+// d, where d represents the state of the system (existing resources) needed for the test.
+func getRunController(t *testing.T, d test.Data) (test.Assets, func()) {
+	t.Helper()
+	names.TestingSeed()
+	return initializeRunControllerAssets(t, d)
+}
+
+// TestReconcile_CloudEvents runs reconcile with a cloud event sink configured
+// to ensure that events are sent in different cases
+func TestReconcile_CloudEvents(t *testing.T) {
+	ignoreResourceVersion := cmpopts.IgnoreFields(v1beta1.CustomRun{}, "ObjectMeta.ResourceVersion")
+
+	cms := []*corev1.ConfigMap{
+		{
+			ObjectMeta: metav1.ObjectMeta{Name: config.GetDefaultsConfigName(), Namespace: system.Namespace()},
+			Data: map[string]string{
+				"default-cloud-events-sink": "http://synk:8080",
+			},
+		}, {
+			ObjectMeta: metav1.ObjectMeta{Name: config.GetFeatureFlagsConfigName(), Namespace: system.Namespace()},
+			Data: map[string]string{
+				"send-cloudevents-for-runs": "true",
+			},
+		},
+	}
+	testcases := []struct {
+		name            string
+		condition       *apis.Condition
+		wantCloudEvents []string
+	}{{
+		name:            "Run with no condition",
+		condition:       nil,
+		wantCloudEvents: []string{`(?s)dev.tekton.event.run.started.v1.*test-run`},
+	}, {
+		name: "Run with unknown condition",
+		condition: &apis.Condition{
+			Type:   apis.ConditionSucceeded,
+			Status: corev1.ConditionUnknown,
+			Reason: v1beta1.TaskRunReasonRunning.String(),
```

Comment:
why are we using `v1beta1.TaskRunReasonRunning`? Is it because we've not defined `v1alpha1.RunReasonRunning` or `v1beta1.CustomRunReasonRunning`?

Reply:
I guess so? =)

Reply:
What do you think of defining CustomRunReasons in https://github.com/tektoncd/pipeline/blob/main/pkg/apis/run/v1beta1/customrunstatus_types.go?

Can have the basic ones:

```go
const (
	// CustomRunReasonStarted is the reason set when the CustomRun has just started
	CustomRunReasonStarted CustomRunReason = "Started"

	// CustomRunReasonRunning is the reason set when the CustomRun is running
	CustomRunReasonRunning CustomRunReason = "Running"

	// CustomRunReasonSuccessful is the reason set when the CustomRun completed successfully
	CustomRunReasonSuccessful CustomRunReason = "Succeeded"

	// CustomRunReasonFailed is the reason set when the CustomRun completed with a failure
	CustomRunReasonFailed CustomRunReason = "Failed"
)
```

Reply:
but these constants are redefined for each type, can we reuse them somehow? 🤔